### PR TITLE
update dependencies

### LINKED ISSUE #71: Update reason & ocaml dependency
The versions of `ocaml` and `reason` in master are behind the current `latest` versions.

Updating to latest version unblocked me when compiling with gcc 10.

### PR DESCRIPTION
**Pull request by woeps**
- `reason` and `ocaml` dependencies updated to current latest version:
  - `reason` to `3.6.0`
  - `ocaml` to `4.10.0`
- Dune lang version updated to `2.6` in `dune-project` file.
- `.ci/utils/use-esy.yml` updated to use esy version `0.6.6`.

closes #71

### CODE REVIEW

File: package.json
Comment (yawaramin) [inline package.json]: I think this can be removed? Since `ocaml` is already a runtime dep.

**Review by yawaramin — APPROVED** [review]
Great, thank you.

**Comment by yawaramin** [comment]
@woeps since this PR is upgrading stuff, would you mind upgrading to Esy 0.6.6 for the CI builds: https://github.com/esy-ocaml/hello-reason/blob/903f528740c6689a3213f167f2d22153e8b89cae/.ci/utils/use-esy.yml

File: package.json
Comment (Et7f3) [inline package.json]:
```suggestion
    "@opam/dune": ">=1.11",
```
because you have updated dune-project.

**Review by Et7f3 — CHANGES_REQUESTED** [review]

**Comment by Et7f3** [comment]
We still use dune-lang (1.0 or 1.11 with this PR) maybe it is time to use (dune-lang 2.6) so new package that use this template have recent dependencies.

**Comment by woeps** [comment]
@Et7f3 I updated dune lang to `2.6` in `dune-project`, should I update dune version in `package.json` also to `2.6`?

File: package.json
Comment (yawaramin) [inline package.json]: Since you set `(lang dune 2.6)`, to avoid any possible issues I would say use version `>= 2.6` here

**Comment by Et7f3** [comment]
Yup they must be the same

**Review by Et7f3 — APPROVED** [review]

**Comment by ulrikstrid** [comment]
Are we using features from fune 2.6 in the dune-project?

**Comment by woeps** [comment]
> Are we using features from fune 2.6 in the dune-project?

currently `dune-project` just contains:
```
(lang dune 2.6)
```

**Review by ulrikstrid — APPROVED** [review]
I'm fine with merging this, but it might make sense to be more relaxed on the dune version (^2.0.0) since we're not using any of the new/advanced dune features.

**Comment by yawaramin** [comment]
@ulrikstrid all else being equal, imho I'd prefer this repo to be as up-to-date as possible. Especially because this repo is oriented toward building apps. Library publishers can be taught to relax their dune version requirements.

**Comment by ulrikstrid** [comment]
What happens when you hit a package with a upper limit on dune @yawaramin?

**Comment by yawaramin** [comment]
@ulrikstrid send a PR to the package to upgrade dune and in the meanwhile downgrade locally ;-)

Seriously though, I set `>= 2.6.2 < 3.0.0` as my dune version in ReWeb and haven't had any issues with it. From what I can tell none of my dependencies set an upper bound on dune.

**Comment by ulrikstrid** [comment]
@jordwalke what's your take? I think you hit some dune constraint issue at some point right?

Again, I'm okay with merging this, but what is the "best practice"?

**Comment by ducaale** [comment]
@woeps if I may ask, how you were able to generate a lockfile that works with OCaml 4.10.0? This PR worked for me in macOS but when I tried cloning this repo and manually re-applying your changes in `package.json` and in `dune`, I got errors because`refmterr` doesn't support OCaml 4.10.0 yet?

```
File "src/reason-parser/reason_parser_def.ml", line 1, characters 5-36:
    1 | open Migrate_parsetree.OCaml_408.Ast
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    Error: Unbound module Migrate_parsetree.OCaml_408
```

My plan is to try updating the dependencies of https://github.com/esy-ocaml/hello-ocaml as it hasn't been updated for some time.

**Comment by yawaramin** [comment]
@ducaale did you upgrade the `reason` version as well? https://github.com/esy-ocaml/hello-reason/pull/72/files#diff-b9cfc7f2cdf78a7f4b91a753d10865a2R28

**Comment by ducaale** [comment]
@yawaramin Yes, I did.

`package.json`
```json
"dependencies": {
  "@opam/dune": ">=2.6",
  "@reason-native/console": "*",
  "@reason-native/pastel": "*",
  "@esy-ocaml/reason": ">= 3.6.0 < 4.0.0",
  "ocaml": "~4.10.0"
},
"devDependencies": {
  "@opam/merlin": "*",
  "@reason-native/rely": "*",
  "refmterr": "*",
  "@opam/odoc": "*"
}
```

`dune-project`
```
(lang dune 2.6)
```

**Comment by yawaramin** [comment]
@woeps @ulrikstrid can we move this forward? I have no issues with using `(lang dune 2.0)`. Does that unblock this?

**Comment by jordwalke** [comment]
If this fixes something feel free to merge, and then we can iterate again in more diffs.

But you should probably delete the lockfile and run esy again because I just pushed out a new version of Reason with much wider omp compatibility (you should get reason@3.6.2).

I have historically used this repo to create both apps and libraries. I think we should be encouraging people to upgrade to the latest dependencies as much as possible, however, libraries should have the widest range of compatibility possible. We can make pesy library templates focus on compatibility of dependencies, and it doesn't matter as much for this specific repo (but if I was forced to vote, I'd pick wider compatibility).

Feel free to land. If you update to reason 3.6.2 it will save me a PR. ;)

**Comment by woeps** [comment]
I will push an update of reason to 3.6.2 tonight (UTC+2).

**Comment by ulrikstrid** [comment]
Thank you!

**Comment by Et7f3** [comment]
dune 2.7 was released just before the merge 😀

**Comment by woeps** [comment]
@ulrikstrid I just realised my last commit updated dune in esy.lock to 2.7 but I forgot to update that version in other places:

- `package.json` still states `>=2.6`
- `dune-project` still states `2.6`.

Is this an issue I should send another pr for?

**Comment by ulrikstrid** [comment]
We would rather want to **relax** the dune version to the lowest possible (2.0 probably) since this template can be used for libraries as well and then we don't want to force 2.7 since they will get that anyway.

**Comment by Et7f3** [comment]
@woeps it is not an issue esy.lock only contain the last known solution but constraints are good.
@ulrikstrid if project put upper bound it will be <3.0.0 so requiring >=2.6.0 or >=2.7.0 should be fine (although for dune if we add a dune-project file the upper bound shouldn't be needed so it is less than a issue)
I have checked the number of download per month:
1.2.1: 1179
1.11.4: 7286
<2.0 >2.5: ~3k
2.5.1: 19351
2.6.0: 4327
2.6.1: 40041
2.6.2: 35307
2.7.0: 1866 (but released yesterday)
So 2.6 is definitely a good standard and why not 2.7 because it is likely that users will update their lockdir soon (or at least in the month)

**Pull request merged**
